### DISCUSSION TITLE
wlink failing to link libraries specified for Win95 project

### DISCUSSION THREAD

**Discussion by BlueFrog76** [open-watcom/open-watcom-v2#1082] [Q&A]
I have been provided a Windows program with Fortran and C++ source (mostly Fortran) that was compiled with a previous version of Watcom (I don't know what previous version).  The Watcom project and target files provided compile all the source files using Open Watcom 2.0 (fortran 77 version), but when it comes time to link the libraries, I get several undefined reference messages:
Error! E2028: _GetTextExtentPointA is an undefined reference
Error! E2028: _SendDlgItemMessageA is an undefined reference
Error! E2028: _WritePrivateProfileStringA is an undefined reference
Error! E2028: _GetTextMetricsA is an undefined reference
Error! E2028: _SendMessageA is an undefined reference
Error! E2028: _MessageBoxA is an undefined reference
Error! E2028: _GetObjectA is an undefined reference
Error! E2028: _TextOutA is an undefined reference
Error! E2028: _CreateFontIndirectA is an undefined reference
Error! E2028: _DialogBoxParamA is an undefined reference
Error! E2028: _PostMessageA is an undefined reference
Error! E2028: _LoadCursorA is an undefined reference
Error! E2028: _GetMessageA is an undefined reference
Error! E2028: _GetForegroundWindow@4 is an undefined reference
Error! E2028: OPENFILEHOOKPROC is an undefined reference
Error! E2028: _GetOpenFileNameA is an undefined reference
Error! E2028: _GetSaveFileNameA is an undefined reference
Error! E2028: _SetDlgItemTextA is an undefined reference
Error! E2028: _GetWindowTextLengthA is an undefined reference
Error! E2028: _GetDlgItemTextA is an undefined reference
Error! E2028: GENMESHDLGPROC is an undefined reference
Error! E2028: _LoadStringA is an undefined reference
Error! E2028: _GetWindowLongA is an undefined reference
Error! E2028: _SetWindowLongA is an undefined reference
Error! E2028: _CallWindowProcA is an undefined reference
Error! E3137: too many errors encountered

The Target includes several libraries and I have a previous map file that identifies the libraries that were included and used.  I'm not sure why the libraries don't seem to be included.  I did attempt to ensure they were added by going to the Options>Windows Linking Switches dialog.  I selected Import, Export, and Library Switches.  I specified the paths to the relevant libraries, but I still get the same errors.

Just before the errors, the compile message in the IDE log is 
wlink name dplot32 sys win95 op m libp c:\watcom\lib386\nt libp c:\watcom\lib386 libp e:\CodeProjects\Dplot\lib libr comctl32 libr comdlg32 libr gdi32 libr kernel32 libr user32 libr fparser libr utils

I've tried variations and just listing one of the libraries at a time, including and leaving off the .lib extensions, etc.  It really seems like there is something being done differently with the linker options that I am not getting right.  Any guidance would be greatly appreciated.

**Comment by jmalak** [comment]
Please put here linker map file, but it looks like some wrong options used by Fortran compiler.
Could you give here your Fortran compiler command line you use.
Anyway it looks like you use non-Watcom or very old libraries or libraries compiled by wery old Watcom version.
I deduce it from unresolved symbols which is not standard.
By example symbol _GetOpenFileNameA should be _GetOpenFileNameA@4. Symbol _GetOpenFileNameA doesn't exist in Windows.
May be wrong Windows Fortran header file for compilation is used, it can be more issues.

**Reply by BlueFrog76** [reply]
[dplot32.map.txt](https://github.com/open-watcom/open-watcom-v2/files/11442383/dplot32.map.txt)
[dplot32-old.map.txt](https://github.com/open-watcom/open-watcom-v2/files/11442391/dplot32-old.map.txt)
[dplot32-log.txt](https://github.com/open-watcom/open-watcom-v2/files/11442496/dplot32-log.txt)
These are the new map and old map outputs and the compiler log from the IDE.

**Reply by jmalak** [reply]
Please, try to use latest build, because a few days before I fixed Fortran compiler issue with non-Fortran symbols.

**Reply by BlueFrog76** [reply]
I have tried the latest build, but I still am having the same issue with the linker.  I have tried changing some of the routines to newer versions, but the linker still seems to fail.  I also looked in the gdi32.lib and some of the other current library files and it looks like at least some of the reported undefined references are identified in the library.  Is there something that I am perhaps not understanding about the linker or compiler settings?  The User Guide seems to just focus on the command line instructions, not the interface.  See the sample input from the IDE.  This produces the following compile and link instruction, which is followed by the list of errors noted above:
![image](https://github.com/open-watcom/open-watcom-v2/assets/6477855/0f119999-0985-405c-88d6-30e47449f6b4)
There are other non-windows libraries that link fine.  So I am wondering if perhaps there is an issue with how I am trying to link to the windows libraries, perhaps some step or instruction I am missing.
"wlink name dplot32 sys nt_win op m libp D:\CodeProjects\DPLOTW\dahs\dplotw\source\lib libp D:\CodeProjects\DPLOTW\dahs\dplotw\source libp C:\wc\lib386 libp C:\wc\lib386\nt libp C:\wc\lib386 libp C:\wc\lib386 libr dplotavi libr fparser libr freeimage libr comctl32 libr comdlg32 libr gdi32 libr kernel32 libr user32 libr utils libr math387r libr emu387 libr flib7 libr clib3r libr advapi32 op maxe=25 op q op symf op el @dplot32.lk1"
[DPLOT32-ARA.txt](https://github.com/open-watcom/open-watcom-v2/files/11479727/DPLOT32-ARA.txt)
[dplot32.map.txt](https://github.com/open-watcom/open-watcom-v2/files/11479730/dplot32.map.txt)

**Reply by jmalak** [reply]
Sorry, what you put here is error listing due to linker option **"op maxe=25"**.
Please remove this linker option then linker will generate full map file not only errors.
![image](https://github.com/open-watcom/open-watcom-v2/assets/2017260/b203bbe0-2ddd-4416-a75b-0b541f26dc0e)
Probably problem is with some library which is incompatible for some reason.
As soon as we get full map file we will see which one.

Anyway your C source files report W308, you should correct it.

I think your installation is strange, because C code refere to standard OW C headers in **C:\wc\h\nt** and folowing reference signal that you have some mixture of OW header files with some unknown header files
C:\wc\h\nt\winpackagefamily.h(91): Warning! W138: No newline at end of file
C:\wc\h\nt\winapifamily.h(234): Warning! W138: No newline at end of file
These two header files don't exists in OW distribution.
What is origin of these header files?
Do you have installed OW C/C++ with OW Fortran?

**Reply by BlueFrog76** [reply]
Yes, I am compiling a Fortran project that also has some C++ files.  I have installed both the OW C/C++ and OW Fortran, both from the current 2.0 release.  I will look into resolving the C warnings.  I received this project after we acquired it from another company.  The program compiled fine on their computer and version of Watcom, but I do not know what version they were using.

They had a link to multimon.h, which was not present.  I found a copy of multimon.h, winpackagefamily.h, and winapifamily.h from microsoft to get past the absence of this file.  I don't think it is needed anymore based on what I am reading, but I'm just trying to figure out how to get the program to compile and link before we start making updates to modernize the code for a future release.
[dplot32.map.txt](https://github.com/open-watcom/open-watcom-v2/files/11481043/dplot32.map.txt)

Here is the expanded map file after I removed the op maxe=25 limit

**Reply by jmalak** [reply]
As expected there is problem with two versions of Win32 libraries.
One set is OW (correct) and second one is in  ../lib, see map file for used libraries.
C:\wc\lib386\nt\kernel32.lib
C:\wc\lib386\nt\user32.lib
C:\wc\lib386\nt\gdi32.lib
C:\wc\lib386\nt\advapi32.lib
C:\wc\lib386\nt\comdlg32.lib
C:\wc\lib386\nt\winspool.lib
C:\wc\lib386\nt\shell32.lib
D:\CodeProjects\DPLOTW\dahs\dplotw\source\dplotavi.lib
D:\CodeProjects\DPLOTW\dahs\dplotw\source\fparser.lib
D:\CodeProjects\DPLOTW\dahs\dplotw\source\freeimage.lib
D:\CodeProjects\DPLOTW\dahs\dplotw\source\utils.lib
C:\wc\lib386\math387r.lib
C:\wc\lib386\emu387.lib
C:\wc\lib386\nt\flib7.lib
C:\wc\lib386\nt\clib3r.lib
..\lib\comctl32.lib
..\lib\comdlg32.lib
..\lib\gdi32.lib
..\lib\kernel32.lib
..\lib\user32.lib

Next problem is your Fortran source code compilation, it looks like missing or incorrect header file for Fortran Win32 API declarations.
OW don't have Fortran Win32 API header files. There is only Windows 3.x API header files (for 16-bit Windows or WIN386 32-bit OW Extender). If you call Win32 API function directly from Fortran code then you need to define appropriate API function as proper Fortran symbols with **`#pragma aux`**.
From map file you see what Win32 API function declarations are missing (Undefined Symbol) and need to be declared as Win32 API.

Below is example (2 API entries) how you could define Win32 API for OW Fortran

```
c$pragma aux (__stdcall) callback
c$pragma aux (__stdcall) winapi
c$pragma aux (__cdecl) windecl

c$pragma aux (winapi) MessageBoxA parm( value*4, value near, value near, value*4 )
c$pragma aux (winapi) SendMessageA parm( value*4, value*4, value*4, value*4 )

      external SendMessageA
      integer*4 SendMessageA

      external MessageBoxA
      integer*4 MessageBoxA

```
it is [link to OW Fortran documentation](https://open-watcom.github.io/open-watcom-v2-wikidocs/fuguide.html#32Mbit__Passing_Arguments_to_nonMFORTRAN_Subprograms) how you define parameters for external functions
